cmd/docker/builder.go

```diff
@@ -0,0 +1,134 @@
+package main
+
+import (
+	"fmt"
+	"os"
+	"strconv"
+
+	pluginmanager "github.com/docker/cli/cli-plugins/manager"
+	"github.com/docker/cli/cli/command"
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+)
+
+const (
+	builderDefaultPlugin = "buildx"
+	buildxMissingWarning = `DEPRECATED: The legacy builder is deprecated and will be removed in a future release.
+            Install the buildx component to build images with BuildKit:
+            https://docs.docker.com/go/buildx/
+`
+
+	buildxMissingError = `ERROR: BuildKit is enabled but the buildx component is missing or broken.
+       Install the buildx component to build images with BuildKit:
+       https://docs.docker.com/go/buildx/
+`
+)
+
+func newBuilderError(warn bool, err error) error {
+	var errorMsg string
+	if warn {
+		errorMsg = buildxMissingWarning
+	} else {
+		errorMsg = buildxMissingError
+	}
+	if pluginmanager.IsNotFound(err) {
+		return errors.New(errorMsg)
+	}
+	if err != nil {
+		return fmt.Errorf("%w\n\n%s", err, errorMsg)
+	}
+	return fmt.Errorf("%s", errorMsg)
+}
+
+func processBuilder(dockerCli command.Cli, cmd *cobra.Command, args, osargs []string) ([]string, []string, error) {
+	var useLegacy bool
+	var useBuilder bool
+
+	// check DOCKER_BUILDKIT env var is present and
+	// if not assume we want to use the builder component
+	if v, ok := os.LookupEnv("DOCKER_BUILDKIT"); ok {
+		enabled, err := strconv.ParseBool(v)
+		if err != nil {
+			return args, osargs, errors.Wrap(err, "DOCKER_BUILDKIT environment variable expects boolean value")
+		}
+		if !enabled {
+			useLegacy = true
+		} else {
+			useBuilder = true
+		}
+	}
+
+	// BuildKit does not yet support Windows, so use classic builder
+	if dockerCli.ServerInfo().OSType == "windows" {
+		return args, osargs, nil
+	}
+
+	// is this a build that should be forwarded to the builder?
+	fwargs, fwosargs, forwarded := forwardBuilder(args, osargs)
+	if !forwarded {
+		return args, osargs, nil
+	}
+
+	if useLegacy {
+		// display warning if not in quiet mode and continue
+		if !isQuietBuild(fwargs, fwosargs) {
+			_, _ = fmt.Fprintln(dockerCli.Err(), newBuilderError(true, nil))
+		}
+		return args, osargs, nil
+	}
+
+	// check plugin is available if cmd forwarded
+	plugin, perr := pluginmanager.GetPlugin(builderDefaultPlugin, dockerCli, cmd.Root())
```

Comment:
Not for this PR, but if I have some time, I want to explore if we can somehow avoid this call, and (if we detected we're going to run a `builder` command), capture the error from `tryPluginRun()` https://github.com/docker/cli/blob/a32cd16160f1b41c1c4ae7bee4dac929d1484e59/cmd/docker/docker.go#L274-L277


That way we could avoid doing the plugin lookup code (which traverses various directories, calls plugin binaries etc etc)